python/lsst/obs/base/mapping.py

```diff
@@ -151,31 +152,68 @@ def lookup(self, properties, dataId):
         if self.registry is None:
             raise RuntimeError("No registry for lookup")
 
+        skyMapKeys = ("tract", "patch")
+
         where = []
         values = []
+
+        # Prepare to remove skymap entries from properties list.  These must
+        # be in the data ID, so we store which ones we're removing and create
+        # an OrderedDict that tells us where to re-insert them.  That maps the
+        # name of the property to either its index in the properties list
+        # *after* the skymap ones have been removed (for entries that aren't
+        # skymap ones) or the value from the data ID (for those that are).
+        removed = set()
+        substitutions = OrderedDict()
+        index = 0
+        properties = list(properties)  # don't modify the original list
+        for p in properties:
+            if p in skyMapKeys:
+                try:
+                    substitutions[p] = dataId[p]
+                    removed.add(p)
+                except KeyError:
+                    raise RuntimeError(
+                        "Cannot look up skymap key '%s'; it must be explicitly included in the data ID" % p
+                    )
+            else:
+                substitutions[p] = index
+                index += 1
+        # Can't actually remove while iterating above, so we do it here.
+        for p in removed:
+            properties.remove(p)
+
         fastPath = True
         for p in properties:
             if p not in ('filter', 'expTime', 'taiObs'):
                 fastPath = False
                 break
         if fastPath and 'visit' in dataId and "raw" in self.tables:
             lookupDataId = {'visit': dataId['visit']}
-            self.registry.lookup(properties, 'raw_visit', lookupDataId)
```

Comment:
It looks like it might have gone from optimization to no-op in e522602a.  The intent was for it to be invoked fairly frequently, but perhaps that also went by the wayside a while ago.